src/Uno.UI/UI/Xaml/Controls/ListViewBase/ListViewBase.cs

```diff
@@ -501,33 +584,141 @@ internal override void OnItemClicked(int clickedIndex)
+					{
+						SelectedIndex = -1;
+					}
+					else
+					{
+						SelectedIndex = clickedIndex;
+					}
+
 					break;
 				case ListViewSelectionMode.Multiple:
+					if (!modifiers.HasFlag(VirtualKeyModifiers.Shift))
+					{
+						FlipSelectionInMultipleSelection(clickedIndex, item: clickedItem, container: clickedContainer);
+						break;
+					}
+
+					// if shift is held, all items between the clicked item and the previously
+					// focused item match the selection of the previously focused item
+					var newSelection = focusedContainer?.IsSelected ?? false;
+
+					var multipleLowerBound = Math.Min(focusedIndex, clickedIndex);
+					var multipleUpperBound = Math.Max(focusedIndex, clickedIndex);
+
+					for (int currentIndex = multipleLowerBound; currentIndex <= multipleUpperBound; currentIndex++)
+					{
+						if (ContainerFromIndex(currentIndex) is SelectorItem container && container.IsSelected != newSelection)
+						{
+							FlipSelectionInMultipleSelection(currentIndex, container: container);
+						}
+					}
+
+					break;
 				case ListViewSelectionMode.Extended:
-					HandleMultipleSelection(clickedIndex, item);
+					if (modifiers.HasFlag(VirtualKeyModifiers.Shift))
+					{
+						// if shift is held, all items between the clicked item and the previously
+						// focused item are selected. Everything else is unselected.
+						var extendedLowerBound = Math.Min(clickedIndex, ExtendedShiftSelectionStart);
+						var extendedUpperBound = Math.Max(clickedIndex, ExtendedShiftSelectionStart);
+
+						for (var currentIndex = 0; currentIndex < Items.Count; currentIndex++)
+						{
+							var currentContainer = ContainerFromIndex(currentIndex) as SelectorItem;
+
+							if (currentContainer == null)
+							{
+								continue;
+							}
+
+							if (currentIndex > extendedUpperBound || currentIndex < extendedLowerBound)
+							{
+								UnselectInMultipleSelection(currentIndex);
+							}
+							else
+							{
+								SelectInMultipleSelection(currentIndex);
+							}
+						}
+
+						break;
+					}
+
+					// any click but a shift-click marks the start of a new selection
+					// even if the click actually unselects an item 
+					ExtendedShiftSelectionStart = clickedIndex;
+
+					if (modifiers.HasFlag(VirtualKeyModifiers.Control))
+					{
+						FlipSelectionInMultipleSelection(clickedIndex, item: clickedItem);
+						break;
+					}
+
+					// no modifiers
+					// Act like single selection mode i.e. set one, unset everything else
+					for (int i = 0; i < Items.Count; i++)
+					{
+						if (i != clickedIndex)
+						{
+							UnselectInMultipleSelection(i);
+						}
+					}
+
+					SelectInMultipleSelection(clickedIndex, item: clickedItem);
+
 					break;
 			}
 		}
 
-		private void HandleMultipleSelection(int clickedIndex, object item)
+		private void FlipSelectionInMultipleSelection(int index, object item = null, SelectorItem container = null)
 		{
+			if (!SelectInMultipleSelection(index, item, container))
```

Comment:
you have access to the `container` already, just branch base on `container.IsSelected`
and, just inline Un/SelectInMultipleSelection here as before